### PR TITLE
Fix calico-node restarts on BIRD mode

### LINKED ISSUE #7191: Calico pod restarts due to livenessProbe failures when using BIRD mode.
Created to track this issue from https://github.com/k0sproject/k0s/issues/6988

As referenced in: https://github.com/k0sproject/k0s/issues/6988#issuecomment-3974145203 when calico is using BIRD mode livenessProbes fail causing the pod to restart

### PR DESCRIPTION
**Pull request by juanluisvaladas**
## Description

Bumps calico to v3.31.4-1 which fixes #7191 and also adds tests that would have caught this issue.

Fixes #7191

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

- [x] Manual test
- [x] Auto test added

## Checklist

- [x] My code follows the style [guidelines](https://docs.k0sproject.io/head/contributors/) of this project
- [x] My commit messages are [signed-off](https://docs.k0sproject.io/head/contributors/github_workflow/)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: inttest/common/util.go [hunk 1]
```diff
@@ -419,3 +419,40 @@ func (s *LineWriter) Flush() {
...
+			}
+		}
+	}
+	pods, err := client.CoreV1().Pods(metav1.NamespaceSystem).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		errs = append(errs, err)
+	}
+	for _, pod := range pods.Items {
+		for _, cs := range pod.Status.ContainerStatuses {
+			if cs.RestartCount != int32(0) {
```
Comment (twz123) [inline inttest/common/util.go:461]: I'm a bit reluctant to add a test for zero pod restarts. Isn't it enough if a Pod is ready?

File: inttest/common/util.go [hunk 2]
```diff
@@ -419,3 +419,40 @@ func (s *LineWriter) Flush() {
 		s.buf = s.buf[:0]
 	}
 }
+
+// VerifyNoRestartedPods returns a slice of errors with the restarted on kube-system.VerifyNoRestartedPods
+// In order to. verify this, the test makes sure that the node has been ready for at least 80 seconds,
+// which is enough to validate that the calico (slowest component to restart due to a liveness probe error)
+// would have restarted.
```
Comment (twz123) [inline inttest/common/util.go]: Some full stops at places where they shouldn't be.

File: inttest/common/util.go [hunk 3]
```diff
@@ -419,3 +419,40 @@ func (s *LineWriter) Flush() {
...
+
+	for _, node := range nodes.Items {
+		for _, condition := range node.Status.Conditions {
+			if condition.Type == corev1.NodeReady {
+				if condition.Status == corev1.ConditionFalse {
+					errs = append(errs, fmt.Errorf("node %s is not ready", node.Name))
+					continue
+				}
+				// Wait at least 80 seconds so that the container mau have rebooted
+				<-time.After(time.Until(condition.LastTransitionTime.Add(80 * time.Second)))
```
Comment (twz123) [inline inttest/common/util.go]: Not sure what we're after, here?

File: inttest/kuberouter/kuberouter_hairpin_test.go [hunk 4]
```diff
@@ -130,6 +135,7 @@ func TestKubeRouterHairpinSuite(t *testing.T) {
 		s.K0sExtraImageBundleMountPoints = []string{"/var/lib/k0s/images/ipv6.tar"}
 	}
 	suite.Run(t, &s)
+
```
Comment (twz123) [inline inttest/kuberouter/kuberouter_hairpin_test.go]: Stray whitespace

File: inttest/common/util.go (hunk 1 above)
Comment (juanluisvaladas) [reply inttest/common/util.go:461]: I don't think we can because calico-node readynessProbe works but livenessProbe fails.

We could limit the pods we search for using labels so that we only validate calico/kube-router pods but if we don't check restarts I can't think of a way to actually validate that this doesn't happen

File: inttest/common/util.go (hunk 2 above)
Comment (juanluisvaladas) [reply inttest/common/util.go]: Fixed it thanks

File: inttest/common/util.go (hunk 3 above)
Comment (juanluisvaladas) [reply inttest/common/util.go]: We want the node to be ready for at least 80 seconds before checking the pods for restarts, otherwise it's possible that the pods haven't had enough time to restart.

Calico needs roughly 70 seconds to fail, but if we test it with `v3.31.4-0` with only 70 seconds the test doesn't fail consistently. At 75 seconds the test fails consistently so I rounded it up to 80 seconds for a bit more reliability.

I added a couple comments in the code hoping that the intent and reason for it is clear now.

Also this isn't respecting the context timeout, I modified it to fix that problem.
